verification/spec/connection-handshake/ConnectionHandshakeModule.tla

```diff
@@ -0,0 +1,353 @@
+    a set of heights.
+
+ ***************************************************************************)
+
+EXTENDS Naturals, FiniteSets, Sequences, ICS3Types
+
+
+CONSTANTS MaxChainHeight,   \* Maximum height of the local chain.
+          ConnectionIDs,    \* The set of valid connection IDs.
+          ClientIDs,        \* The set of valid client IDs.
+          MaxBufLen         \* Maximum length of the input and output buffers.
+
+
+ASSUME Cardinality(ConnectionIDs) >= 1
+ASSUME Cardinality(ClientIDs) >= 1
+
+
+VARIABLES
+(******************************* Store *****************************
+     The store record of a chain contains the following fields:
+     
+     - id -- a string 
+       Stores the identifier of the chain where this module executes.
+
+     - latestHeight -- a Nat
+       Describes the current height of the chain.  
+
+     - connection -- a connection record.
+       Captures all the details of the connection on this chain.
+       For a full description of a connection record, see the
+       'Environment.Connections' set.
+
+     - client -- a client record.
+       Specifies the state of the client running on this chain.
+       For a full description of a client and the initialization values
+       for this record, see the InitClients set below.  
+     
+ ***************************************************************************)   
+    store,
+    (* A buffer (Sequence) holding any message(s) incoming to this module. *)
+    inBuf,
+    (* A buffer (Sequence) holding outbound message(s) from this module. *)
+    outBuf
+
+
+moduleVars == <<inBuf, outBuf, store>>
+
+
+(***************************************************************************
+    Helper operators.
+ ***************************************************************************)
+
+
+(* Returns true if 'para' matches the parameters in the local connection,
+    and returns false otherwise.
+ *)
+CheckLocalParameters(para) ==
+    LET local == store.connection.parameters.localEnd
+        remote == store.connection.parameters.remoteEnd 
+    IN /\ local.connectionID = para.localEnd.connectionID   
+       /\ remote.connectionID = para.remoteEnd.connectionID
+       /\ local.clientID = para.localEnd.clientID       
+       /\ remote.clientID = para.remoteEnd.clientID
+
+
+(* Validates a ConnectionParameter.
+    Expects as input a ConnectionParameter 'para' and returns true or false.
+    
+    This is a basic validation step, making sure that 'para' is valid with
+    respect to module-level constants ConnectionIDs and ClientIDs.
+    If there is a connection in the store, this also validates 'para'
+    against the parameters of an existing connection by relying on the
+    state predicate CheckLocalParameters.
+*)
+ValidConnectionParameters(para) ==
+    /\ para.localEnd.connectionID \in ConnectionIDs
+    /\ para.localEnd.clientID \in ClientIDs
+    /\ \/ store.connection = NullConnection
+       \/ /\ store.connection /= NullConnection
```

Comment:
It is a bit mixing of concerns; maybe this check can be taken out and have this function only validates if ids are valid.